File: src/main/java/gg/grounds/infrastructure/event/StatusEventBus.java
package gg.grounds.infrastructure.event;

import gg.grounds.grpc.status.StatusEvent;
import io.smallrye.mutiny.Multi;
import io.smallrye.mutiny.operators.multi.processors.BroadcastProcessor;
import jakarta.enterprise.context.ApplicationScoped;

@ApplicationScoped
public class StatusEventBus {

    private final BroadcastProcessor<StatusEvent> processor = BroadcastProcessor.create();

    public void publish(StatusEvent event) {
        processor.onNext(event);
    }

    public Multi<StatusEvent> stream() {
        return processor;
    }
}

File: src/main/kotlin/gg/grounds/infrastructure/event/StatusEventBus.kt
package gg.grounds.infrastructure.event

import gg.grounds.grpc.status.StatusEvent
import io.smallrye.mutiny.Multi
import io.smallrye.mutiny.operators.multi.processors.BroadcastProcessor
import jakarta.enterprise.context.ApplicationScoped

@ApplicationScoped
class StatusEventBus {
    private val processor: BroadcastProcessor<StatusEvent?> = BroadcastProcessor.create<StatusEvent?>()

    fun publish(event: StatusEvent?) {
        processor.onNext(event)
    }

    fun stream(): Multi<StatusEvent?> {
        return processor
    }
}
